### ISSUE TITLE
Support slicing of sparse domains

### ISSUE THREAD

**Issue by buddha314** [chapel-lang/chapel#8023]
I'm not sure how much overlap there with [this issue](https://github.com/chapel-lang/chapel/issues/7374), but it would be very nice to be able to traverse rows or columns (and sets thereof) in sparse matrices.

```chapel
var D: domain = [1..5];
var SD: sparse subdomain(D);
var A: [SD] real;
SD += (..);

// maybe something like
for c in A[3,..] {
  writeln("col ", c, " is ", A[3,c]);
}
```

Good things to implement on top of this are
* Number of non-zero entries in that row
* reduce intents like make, min, sum on that row
* Assignment of values to that row

**Comment by buddha314** [comment]
In the mean time, I wrote this terrible hack to get me the actual neighbors. Is there an intermediate solution that won't be so slow?

```chapel
  proc Graph.neighbors(vid: int) {
    var result: [1..0] int;
    for (i,j) in this.A.domain {
      if i == vid {
        result.push_back(j);
      }
    }
    return result;
  }
```

**Comment by bradcray** [comment]
You're interested in the neighbors of a given row, right?  I'd do it as follows:

```chapel
use LayoutCS;

const D = {1..4, 1..4};
var SD: sparse subdomain(D) dmapped CS();
SD += (1,2);
SD += (2,1);
SD += (3,3);
SD += (3,4);
SD += (4,1);
SD += (4,4);

iter neighbors(SparseDom, row) {
  for col in SparseDom.dimIter(2, row) do
    yield col;
}

config const row = 3;

for col	in neighbors(SD, row) do
  writeln(col, " is a neighbor of ", row);
```

**Comment by buddha314** [comment]
Is there a cool method like that for the diagonal?  I don't see how you would use `.dimIter`.  I want to set the diagonal of a sparse matrix to 0.

**Comment by buddha314** [comment]
And how do I put an `if` in there, like
```chapel
iter neighbors(SparseDom, row) {
  for col in SparseDom.dimIter(2, row) do
    if col != row yield col;
}```

E.g., I want to eliminate the diagonal.  You may be sensing a theme.

**Comment by bradcray** [comment]
Putting the `if` in there just like that ought to work.  Does it not?

**Comment by buddha314** [comment]
Yes, like on [Stack Overflow](https://stackoverflow.com/questions/47823225/chapel-iterators-with-conditionals/47823373#47823373)  When you're right, you're right.  And you, my friend, are right.

**Comment by buddha314** [comment]
I know I'm sneaking around the undocumented bits here.  I used this to iterate over rows in a sparse matrix, `vid` is the row ID I'm jamming along.

```chapel
for col in this.SD.dimIter(2,vid) do if col != vid then result.push_back(col);
```

And it seems to work fine. But then I tried to iterate over columns with

```chapel
for row in this.SD.dimIter(vid,2) do if row != vid then result.push_back(row);
for row in this.SD.dimIter(3,vid) do if row != vid then result.push_back(row);
```

and a couple others to no avail.  Any thoughts on that?

**Comment by bradcray** [comment]
@buddha314 : This relates back to the issue #7374 you referred to at the outset of this issue.  The way you would go over columns would be `.dimIter(1, vid)` ("iterate along the first dimension at column `vid`).  That said, for a CS domain, we only support iteration over the "fast" dimension (rows for CSR, cols for CSC) because iteration over the other tends to be way too expensive.  @e-kayrakli's `dsiPartialThese()` that [he mentions](https://github.com/chapel-lang/chapel/issues/7374#issuecomment-330648498) in that same issue may be your saving grace if you want to do it "expense be damned!"

**Comment by buddha314** [comment]
Following up on your comment, if I have a matrix defined as

```chapel
  var D: domain(2),
      SD = CSRDomain(D),
      X: [SD] real,  // the actual data
```

I want the whole row as an array, can I do 

```chapel
for row in X.dimiter(1, ?)  {
  const r = row
}
```

or something similar?  Perhaps there is a cleaner way?

**Comment by bradcray** [comment]
@buddha314: Sorry for the delayed response.  To capture a row in a (dense) array, I'd do one of the following, depending on what you mean by "as an array":

If you wanted to store the nonzeroes of a row in a dense, packed array (so that if there were 3 nonzeroes in a row, there'd be 3 entries in the array, you could do:

```chapel
forall row in 1..10 {
  var rowData: [1..0] real;
  for col in SD.dimIter(2, row) {
    rowData.push_back(X[row, col]);
  }
  writeln("Row ", row, " vals = ", rowData);
}
```

If you want to store it as an array the size of the row rather than the number of nonzeroes, you could do:

```chapel
forall row in 1..10 {
  var rowData: [1..10] real;
  for col in SD.dimIter(2, row) {
    rowData[col] = X[row, col];
  }
  writeln("Row ", row, " vals = ", rowData);
}
```

Let me know if I've misunderstood your question.

**Comment by buddha314** [comment]
In my example, `SD` is a sparse domain.  So we may be talking past each other.

**Comment by bradcray** [comment]
`SD` is a sparse domain in my code as well, and `X` is a sparse array over it.  I'm realizing, though, that in testing my code, I explicitly set `D` to `{1..10, 1..10}` and then hard-coded some bounds into my code which was a mistake.  I've just edited it to remove those assumptions so that it should work with any CSR sparse `SD`.

**Comment by ben-albrecht** [comment]
Some observations shared by @benharsh in another issue:

- We need a 'stridable' method on sparse domains/arrays to avoid issues with shared-code for rectangular slicing (easily achieved by adding a method to BaseSparse*)
- We currently always return a rectangular domain when slicing any domain. We should return a domain of the kind being sliced, or allow the domain map author to decide what that means.
- ArrayViewRankChange and ArrayViewReindex are not set up to handle non-rectangular domains/arrays, and easily break.
- ArrayViewSlice can be made to work relatively quickly, but slices will be printed as if they were rectangular arrays. I think we may want a general way of telling an array to print itself with a given set of indices.
